qiskit/circuit/library/boolean_logic.py

```diff
@@ -21,41 +21,46 @@
 import numpy as np
 from qiskit.circuit import QuantumRegister, QuantumCircuit
 from qiskit.circuit.exceptions import CircuitError
+from qiskit.util import deprecate_arguments
 
 
 class Permutation(QuantumCircuit):
     """An n_qubit circuit that permutes qubits."""
 
+    @deprecate_arguments({'n_qubits': 'num_qubits'})
     def __init__(self,
-                 n_qubits: int,
+                 num_qubits: int,
                  pattern: Optional[List[int]] = None,
-                 seed: Optional[int] = None) -> QuantumCircuit:
+                 seed: Optional[int] = None,
+                 *, n_qubits: Optional[int] = None  # pylint:disable=unused-argument
```

Comment:
Could you please help me understand the pattern I see here and its purpose? I see it repeated a number of times when used with `deprecate_arguments` but not everywhere.

Reply:
It's an additional safeguard as far as I understand, to prevent accidental setting of `num_qubits` and `n_qubits` as unnamed args. 
E.g. `Permutation(num_qubits, pattern, seed, num_qubits)` would raise an error since `deprecate_arguments` maps `n_qubits` (the last argument) to `num_qubits` which has already been provided (first argument) and raises. This might occur in a test case when the number of arguments is inferred via the signature. Pretty much an edge-case though and we can probably safely remove it :slightly_smiling_face: